### PR TITLE
UI Thread Freezing Fix

### LINKED ISSUE #20: Generating triangulation on non UI thread
Trianglify uses UI thread to generate views, this freezes the UI when view is being rendered. It should be  done on seperate thread to avoid freezing up of UI.

### PR DESCRIPTION
**Pull request by arihant-001**
This PR fixes #20 .
It also introduces a new method generateAndInvalidate() in TrianglifyView class to generate Triangulation according to new parameters and then invalidate  it .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,7 +78,7 @@ trianglifyView.setGridWidth(trianglifyView.getWidth())
     app:fillTriangle="true" />
 ```
 
-> TrianglifyView should be invalidated on change of any of the parameter for changes to take effect.  
+> TrianglifyView should be invalidated on change of any of the parameter for changes to take effect.For invalidate TrianglifyView call generateAndInvalidate() method of TrianglifyView class.
```
Comment (suyashmahar) [inline README.md]: .... to take effect. **To** invalidate TrianglifyView...

File: app/src/main/java/com/sdsmdg/kd/trianglifyexample/MainActivity.java [hunk 2]
```diff
@@ -195,6 +183,11 @@ public boolean onOptionsItemSelected(MenuItem item) {
         return true;
     }
 
+    @Override
+    protected void onResume() {
+        super.onResume();
```
Comment (suyashmahar) [inline app/src/main/java/com/sdsmdg/kd/trianglifyexample/MainActivity.java]: This method is redundant.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 3]
```diff
@@ -216,6 +220,13 @@ void generate() {
         this.triangulation = presenter.getSoup();
     }
 
+    public void generateAndInvalidate() {
+        if(generatorTask!=null)
+            generatorTask.cancel(true);
```
Comment (suyashmahar) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: wrap `generatorTask.cancel(true);` around curly brackets. Even if a `if` contains a single statement it should have curly brackets.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 4]
```diff
@@ -216,6 +220,13 @@ void generate() {
...
+        generatorTask = null;
```
Comment (suyashmahar) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: Line 226 is redundant.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 5]
```diff
@@ -262,4 +273,17 @@ public void drawTriangle(Canvas canvas, Triangle2D triangle2D) {
...
+        protected Triangulation doInBackground(TrianglifyView... params) {
+            return params[0].presenter.getSoup();
+        }
+
+        @Override
+        protected void onPostExecute(Triangulation triangulation) {
+            super.onPostExecute(triangulation);
+            setTriangulation(triangulation);
+        }
+    }
```
Comment (suyashmahar) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: Move `TriangleGeneratorTask` and related non UI thread generation logic to presenter

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 6]
```diff
@@ -169,6 +164,7 @@ public Triangulation getTriangulation() {
 
     public void setTriangulation(Triangulation triangulation) {
         this.triangulation = triangulation;
+        invalidate();
```
Comment (dakshitagrawal) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: there should be no command to invalidate the view, we are just setting the triangulation with this method.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 7]
```diff
@@ -216,6 +220,13 @@ void generate() {
         this.triangulation = presenter.getSoup();
     }
 
+    public void generateAndInvalidate() {
+        if(generatorTask!=null)
+            generatorTask.cancel(true);
+        generatorTask = null;
+        generatorTask = new TriangleGeneratorTask();
+        generatorTask.execute(this);
```
Comment (dakshitagrawal) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: you can add plotOnCanvas method here as the execute method ensures that the doInBackground() and onPostExecute() are executed

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java (hunk 6 above)
Comment (suyashmahar) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: Also setTriangulation should return `this` the instance of  TrianglifyView.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 8]
```diff
@@ -204,9 +200,17 @@ public TrianglifyView setRandomColoring(boolean randomColoring) {
     @Override
     protected void onDraw(Canvas canvas) {
         super.onDraw(canvas);
-        generateAndPlot(canvas);
```
Comment (dakshitagrawal) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:202]: We can put the whole process in generateAndPlot() to maintain readability.

Comment (arihant-001) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:202]: I am shifting worker's thread in presenter class

**Comment by suyashmahar** [comment]
@dakshitagrawal97 @ArihantJain456 This PR changes APIs so care should be taken before merging it.

File: README.md (hunk 1 above)
Comment (arihant-001) [reply README.md]: yes it is a mistake

File: app/src/main/java/com/sdsmdg/kd/trianglifyexample/MainActivity.java (hunk 2 above)
Comment (arihant-001) [reply app/src/main/java/com/sdsmdg/kd/trianglifyexample/MainActivity.java]: ok

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java (hunk 3 above)
Comment (arihant-001) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: no problem in either way

Comment (dakshitagrawal) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: better readability

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java (hunk 6 above)
Comment (arihant-001) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: ok

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java (hunk 3 above)
Comment (suyashmahar) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: +it is easier to maintain code with braces, similar goes with comments it is better to write single lined comments with `/*...*/` than `//`

Comment (arihant-001) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: ok

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java (hunk 8 above)
Comment (arihant-001) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:202]: I don't think that it will increase readability

**Comment by kdexd** [comment]
@ArihantJain456 you should configure your git config properly - either your name or email is incorrect on your local machine.

File: README.md [hunk 9]
```diff
@@ -78,7 +78,7 @@ trianglifyView.setGridWidth(trianglifyView.getWidth())
     app:fillTriangle="true" />
 ```
 
-> TrianglifyView should be invalidated on change of any of the parameter for changes to take effect.  
+> TrianglifyView should be invalidated on change of any of the parameter for changes to take effect.To invalidate TrianglifyView call generateAndInvalidate() method of TrianglifyView class.
```
Comment (kdexd) [inline README.md]: " call `generateAndInvalidate` "

Comment (arihant-001) [reply README.md]: by using () user will understand better

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 10]
```diff
@@ -108,4 +112,29 @@ private Triangulation generateColoredSoup(Triangulation inputTriangulation) {
...
+        if(generatorTask!=null) {
+            generatorTask.cancel(true);
+        }
+        generatorTask = new TriangleGeneratorTask();
+        generatorTask.execute();    }
+
+
+
+
+
```
Comment (kdexd) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: reduce to two blank lines.

File: README.md (hunk 9 above)
Comment (kdexd) [reply README.md]: I haven't seen parenthesis in documentations, and I think "call" word makes it clear. besides our "user' is actually an Android Developer who would be habituated at looking API documentation :wink:

Comment (arihant-001) [reply README.md]: ok

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 11]
```diff
@@ -108,4 +112,29 @@ private Triangulation generateColoredSoup(Triangulation inputTriangulation) {
...
+        @Override
+        protected Triangulation doInBackground(Void... params) {
+            return getSoup();
+        }
+
+        @Override
+        protected void onPostExecute(Triangulation triangulation) {
+            super.onPostExecute(triangulation);
+            ((TrianglifyView)view).setTriangulation(triangulation);
+            ((TrianglifyView)view).invalidate();
```
Comment (kdexd) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: Why is the type casting necessary ?

Comment (arihant-001) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: setTriangulation and invalidate are methods of TrianglifyView  class

**Review by kdexd — COMMENTED** [review]
Overall the work here looks great,  you just need to modify your code as per general conventions followed, perform a small lint.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java (hunk 11 above)
Comment (suyashmahar) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: @ArihantJain456, @karandesai-96  is right you shouldn't type cast view to TrianglifyView. Presenter should use methods defined in `com.sdsmdg.kd.trianglify.views.TrianglifyViewInterface` which are independent of implementation (eg. developing trianglify as desktop application). Modify Interface to include methods that are independent of implementation.

File: README.md [hunk 12]
```diff
@@ -78,7 +78,7 @@ trianglifyView.setGridWidth(trianglifyView.getWidth())
     app:fillTriangle="true" />
 ```
 
-> TrianglifyView should be invalidated on change of any of the parameter for changes to take effect.  
+> TrianglifyView should be invalidated on change of any of the parameter for changes to take effect.To invalidate TrianglifyView call generateAndInvalidate method of TrianglifyView class.
```
Comment (kdexd) [inline README.md:82]: `generateAndInvalidate` <- use backticks for highlighting code block.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 13]
```diff
@@ -108,4 +112,25 @@ private Triangulation generateColoredSoup(Triangulation inputTriangulation) {
     public void clearSoup() {
         triangulation = null;
     }
+    public void generateSoupAndInvalidateView(){
```
Comment (kdexd) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: Missed a blank line before this.

**Comment by karandesai-96** [comment]
Make sure to rebase this branch to `develop` before merging. :smile: 
There must have been a PR merge after this PR was opened, which now causes conflicts.

**Comment by kdexd** [comment]
I thought close and reopen would trigger a Travis build but it didn't. Apparently because of the merge conflicts.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 14]
```diff
@@ -10,6 +12,7 @@
 import com.sdsmdg.kd.trianglify.utilities.patterns.Rectangle;
 import com.sdsmdg.kd.trianglify.utilities.colorizers.Colorizer;
 import com.sdsmdg.kd.trianglify.utilities.colorizers.FixedPointsColorizer;
+import com.sdsmdg.kd.trianglify.views.TrianglifyView;
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: \[Checkstyle\] ERROR: Unused import \- com\.sdsmdg\.kd\.trianglify\.views\.TrianglifyView\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 15]
```diff
@@ -28,6 +31,7 @@
 public class Presenter {
     TrianglifyViewInterface view;
     Triangulation triangulation;
+    TriangleGeneratorTask generatorTask;
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: \[Checkstyle\] ERROR: Variable 'generatorTask' must be private and have accessor methods\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 16]
```diff
@@ -108,4 +112,26 @@ private Triangulation generateColoredSoup(Triangulation inputTriangulation) {
     public void clearSoup() {
         triangulation = null;
     }
+
+    public void generateSoupAndInvalidateView() {
+        if (generatorTask!=null) {
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: \[Checkstyle\] ERROR: '\!=' is not preceded with whitespace\.

Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: \[Checkstyle\] ERROR: '\!=' is not followed by whitespace\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java [hunk 17]
```diff
@@ -108,4 +112,26 @@ private Triangulation generateColoredSoup(Triangulation inputTriangulation) {
...
+            generatorTask.cancel(true);
+        }
+        generatorTask = new TriangleGeneratorTask();
+        generatorTask.execute();
+    }
+
+    class TriangleGeneratorTask extends AsyncTask<Void,Void,Triangulation> {
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/presenters/Presenter.java]: \[Checkstyle\] ERROR: ',' is not followed by whitespace\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 18]
```diff
@@ -199,16 +205,18 @@ public TrianglifyView setRandomColoring(boolean randomColoring) {
...
-    }
-
-    void generateAndPlot(Canvas canvas) {
-        generate();
-        plotOnCanvas(canvas);
+        gridHeight = getHeight();
+        gridWidth = getWidth();
+        if(this.triangulation!=null) {
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: \[Checkstyle\] ERROR: 'if' is not followed by whitespace\.

Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: \[Checkstyle\] ERROR: '\!=' is not preceded with whitespace\.

Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: \[Checkstyle\] ERROR: '\!=' is not followed by whitespace\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 19]
```diff
@@ -199,16 +205,18 @@ public TrianglifyView setRandomColoring(boolean randomColoring) {
...
+            plotOnCanvas(canvas);
+        }
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: \[Checkstyle\] ERROR: '\}' at column 9 should be on the same line as the next part of a multi\-block statement\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 20]
```diff
@@ -199,16 +205,18 @@ public TrianglifyView setRandomColoring(boolean randomColoring) {
...
+        else {
+            generateAndInvalidate();
+        }
     }
 
-    void generate() {
-        this.triangulation = presenter.getSoup();
+    public void generateAndInvalidate(){
```
Comment (sputnikci) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/utilities/colorizers/FixedPointsColorizer.java [hunk 21]
```diff
@@ -123,7 +123,7 @@ public Triangulation getColororedTriangulation(){
 
     private int getColorForPoint(Vector2D point){
         if (randomColoring){
-            return colorPalette.getColor(random.nextInt(9)+1);
+            return colorPalette.getColor(random.nextInt(9));
```
Comment (karandesai-96) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/utilities/colorizers/FixedPointsColorizer.java]: Yaha upar space nahi hai

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 22]
```diff
@@ -48,7 +48,7 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
         gridHeight =height;
     }
 
-    public void attributeSetter(TypedArray typedArray){
+    public void attributeSetter(TypedArray typedArray) {
```
Comment (karandesai-96) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java]: Par yaha space rakhte hai. Kyu karte ho yaar aisa ? :P

**Comment by kdexd** [comment]
I love sputnik :stuck_out_tongue: does half of my work

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 23]
```diff
@@ -162,8 +161,15 @@ public Triangulation getTriangulation() {
...
-    public void setTriangulation(Triangulation triangulation) {
+    @Override
+    public void invalidateView(Triangulation triangulation) {
+        this.setTriangulation(triangulation);
+        invalidate();
+    }
+
+    public TrianglifyView setTriangulation(Triangulation triangulation) {
         this.triangulation = triangulation;
+        return this;
```
Comment (dakshitagrawal) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:172]: There are no return types in setter methods.

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java [hunk 24]
```diff
@@ -199,20 +205,21 @@ public TrianglifyView setRandomColoring(boolean randomColoring) {
...
         super.onDraw(canvas);
-        generateAndPlot(canvas);
-    }
-
-    void generateAndPlot(Canvas canvas) {
-        generate();
-        plotOnCanvas(canvas);
+        gridHeight = getHeight();
+        gridWidth = getWidth();
+        if (this.triangulation != null) {
```
Comment (dakshitagrawal) [inline trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:210]: What is the purpose of gridHeight and gridWidth here?

File: trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java (hunk 23 above)
Comment (suyashmahar) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:172]: Returning current instance, allows chaining of setters like:
```java
trianglifyView.set...()
                .set....()
                .set....()
                .set....();
```

Comment (dakshitagrawal) [reply trianglify/src/main/java/com/sdsmdg/kd/trianglify/views/TrianglifyView.java:172]: ok

**Pull request merged**